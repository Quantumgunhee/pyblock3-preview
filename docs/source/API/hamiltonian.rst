pyblock3.hamiltonian
====================

pyblock3.hamiltonian
--------------------

.. automodule:: pyblock3.hamiltonian
   :members:
   :undoc-members:
   :show-inheritance:

pyblock3.fcidump
----------------

.. automodule:: pyblock3.fcidump
   :members:
   :undoc-members:
   :show-inheritance:
