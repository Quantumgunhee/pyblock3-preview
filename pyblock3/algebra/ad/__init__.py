
ENABLE_JAX = False
ENABLE_EINSUM = True