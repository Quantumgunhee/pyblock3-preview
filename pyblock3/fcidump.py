
import numpy as np

PointGroup = {
    "d2h": [8, 0, 7, 6, 1, 5, 2, 3, 4],
    "c2v": [8, 0, 2, 3, 1],
    "c2h": [8, 0, 2, 3, 1],
    "d2": [8, 0, 3, 2, 1],
    "cs": [8, 0, 1],
    "c2": [8, 0, 1],
    "ci": [8, 0, 1],
    "c1": [8, 0]
}


class FCIDUMP:

    def __init__(self, pg='c1', n_sites=0, n_elec=0, twos=0, ipg=0, uhf=False,
                 h1e=None, g2e=None, orb_sym=None, const_e=0):
        self.pg = pg
        self.n_sites = n_sites
        self.n_elec = n_elec
        self.twos = twos
        self.ipg = ipg
        self.h1e = h1e
        self.g2e = g2e  # aa, ab, bb
        self.const_e = const_e
        self.uhf = uhf
        self.general = False
        self.orb_sym = [0] * self.n_sites if orb_sym is None else orb_sym

    def t(self, s, i, j):
        return self.h1e[s][i, j] if self.uhf else self.h1e[i, j]

    def v(self, sij, skl, i, j, k, l):
        if self.uhf:
            if sij == skl:
                return self.g2e[sij + sij][i, j, k, l]
            elif sij == 0 and skl == 1:
                return self.g2e[1][i, j, k, l]
            else:
                return self.g2e[1][k, l, i, j]
        else:
            return self.g2e[i, j, k, l]

    def read(self, filename):
        """
        Read FCI options and integrals from FCIDUMP file.

        Args:
            filename : str
        """
        with open(filename, 'r') as f:
            ff = f.read().lower()
            if '/' in ff:
                pars, ints = ff.split('/')
            elif '&end' in ff:
                pars, ints = ff.split('&end')

        cont = ','.join(pars.split()[1:])
        cont = cont.split(',')
        cont_dict = {}
        p_key = None
        for c in cont:
            if '=' in c or p_key is None:
                p_key, b = c.split('=')
                cont_dict[p_key.strip().lower()] = b.strip()
            elif len(c.strip()) != 0:
                if len(cont_dict[p_key.strip().lower()]) != 0:
                    cont_dict[p_key.strip().lower()] += ',' + c.strip()
                else:
                    cont_dict[p_key.strip().lower()] = c.strip()

        for k, v in cont_dict.items():
            if ',' in v:
                v = cont_dict[k] = v.split(',')

        self.n_sites = int(cont_dict.get('norb'))
        self.twos = int(cont_dict.get('ms2', 0))
        self.ipg = PointGroup[self.pg][int(cont_dict.get('isym', 0))]
        self.n_elec = int(cont_dict.get('nelec', 0))
        self.uhf = int(cont_dict.get('iuhf', 0)) != 0
        self.general = int(cont_dict.get('igeneral', 0)) != 0
        self.orb_sym = [PointGroup[self.pg]
                        [int(i)] for i in cont_dict.get('orbsym')]

        data = []
        for l in ints.split('\n'):
            ll = l.strip()
            if len(ll) == 0 or ll.strip()[0] == '!':
                continue
            ll = ll.split()
            d = float(ll[0])
            i, j, k, l = [int(x) for x in ll[1:]]
            data.append((i, j, k, l, d))
        if not self.uhf:
            self.h1e = np.zeros((self.n_sites, self.n_sites))
            self.g2e = np.zeros((self.n_sites, self.n_sites,
                                 self.n_sites, self.n_sites))
            for i, j, k, l, d in data:
                if i + j + k + l == 0:
                    self.const_e = d
                elif k + l == 0:
                    self.h1e[i - 1, j - 1] = self.h1e[j - 1, i - 1] = d
                else:
                    self.g2e[i - 1, j - 1, k - 1, l - 1] = d
                    if not self.general:
                        self.g2e[j - 1, i - 1, k - 1, l - 1] = d
                        self.g2e[j - 1, i - 1, l - 1, k - 1] = d
                        self.g2e[i - 1, j - 1, l - 1, k - 1] = d
                        self.g2e[k - 1, l - 1, i - 1, j - 1] = d
                        self.g2e[k - 1, l - 1, j - 1, i - 1] = d
                        self.g2e[l - 1, k - 1, j - 1, i - 1] = d
                        self.g2e[l - 1, k - 1, i - 1, j - 1] = d
        else:
            self.h1e = (None, None)
            self.h1e[0] = np.zeros((self.n_sites, self.n_sites))
            self.h1e[1] = np.zeros((self.n_sites, self.n_sites))
            self.g2e = (None, None, None)  # aa, ab, bb
            self.g2e[0] = np.zeros(
                (self.n_sites, self.n_sites, self.n_sites, self.n_sites))
            self.g2e[1] = np.zeros(
                (self.n_sites, self.n_sites, self.n_sites, self.n_sites))
            self.g2e[2] = np.zeros(
                (self.n_sites, self.n_sites, self.n_sites, self.n_sites))
            ip = 0
            for i, j, k, l, d in data:
                if i + j + k + l == 0:
                    ip += 1
                    if ip == 6:
                        self.const_e = d
                elif k + l == 0:
                    assert ip == 3 or ip == 4
                    if ip == 3:
                        self.h1e[0][i - 1, j -
                                    1] = self.h1e[0][j - 1, i - 1] = d
                    elif ip == 4:
                        self.h1e[1][i - 1, j -
                                    1] = self.h1e[1][j - 1, i - 1] = d
                else:
                    ig = [0, 3, 1][ip]
                    self.g2e[ig][i - 1, j - 1, k - 1, l - 1] = d
                    if not self.general:
                        self.g2e[ig][j - 1, i - 1, k - 1, l - 1] = d
                        self.g2e[ig][j - 1, i - 1, l - 1, k - 1] = d
                        self.g2e[ig][i - 1, j - 1, l - 1, k - 1] = d
                    if not self.general and (ip == 0 or ip == 1):
                        self.g2e[ig][k - 1, l - 1, i - 1, j - 1] = d
                        self.g2e[ig][k - 1, l - 1, j - 1, i - 1] = d
                        self.g2e[ig][l - 1, k - 1, j - 1, i - 1] = d
                        self.g2e[ig][l - 1, k - 1, i - 1, j - 1] = d
        return self
