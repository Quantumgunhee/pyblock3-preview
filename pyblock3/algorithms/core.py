
import numpy as np
import pyblock3.algebra.funcs as pbalg
from enum import Enum, auto


def fmt_size(i, suffix='B'):
    if i < 1000:
        return "%d %s" % (i, suffix)
    else:
        a = 1024
        for pf in "KMGTPEZY":
            p = 2
            for k in [10, 100, 1000]:
                if i < k * a:
                    return "%%.%df %%s%%s" % p % (i / a, pf, suffix)
                p -= 1
            a *= 1024
    return "??? " + suffix

class DecompositionTypes(Enum):
    DensityMatrix = auto()
    SVD = auto()


class NoiseTypes(Enum):
    Perturbative = auto()
    Random = auto()


class SweepAlgorithm:
    def __init__(self, cutoff=1E-14,
                 decomp_type=DecompositionTypes.DensityMatrix,
                 noise_type=NoiseTypes.Perturbative):
        self.cutoff = cutoff
        self.decomp_type = decomp_type
        self.noise_type = noise_type

    def add_dm_noise(self, dm, mpo, wfn, noise, forward):
        if noise == 0:
            return dm
        if self.noise_type == NoiseTypes.Random:
            pdm = dm.__class__.random_like(dm)
            return dm + (noise / np.linalg.norm(pdm)) * pdm
        elif self.noise_type == NoiseTypes.Perturbative:
            if forward:
                pket = np.tensordot(
                    wfn[0], mpo[0], axes=((1, 2), (3, 4)))
                pket.normalize_along_axis(axis=7)
                pdm = np.tensordot(pket, pket, axes=((1, 2, 3, 4, 7), ) * 2)
            else:
                pket = np.tensordot(
                    mpo[1], wfn[0], axes=((3, 4), (3, 4)))
                pket.normalize_along_axis(axis=0)
                pdm = np.tensordot(pket, pket, axes=((0, 3, 4, 5, 6), ) * 2)
            norm = np.linalg.norm(pdm)
            return dm + (noise / norm) * pdm if abs(norm) > 1E-12 else dm
        else:
            assert False

    def add_wfn_noise(self, wfn, noise, forward):
        if noise == 0:
            return wfn
        if self.noise_type == NoiseTypes.Random:
            pwfn = wfn.__class__.random_like(wfn)
            return wfn + (np.sqrt(noise) / np.linalg.norm(pwfn)) * pwfn
        else:
            assert False

    def decomp_two_site(self, mpo, wfns, forward, noise, bond_dim, weights=None):
        wfn = wfns[0] if isinstance(wfns, list) else wfns
        if weights is None:
            weights = [1.0 / len(wfns)] * len(wfns)
        if self.decomp_type == DecompositionTypes.DensityMatrix:
            if forward:
                dm = np.tensordot(
                    wfn[0], wfn[0], axes=((-3, -2, -1), ) * 2)
                if isinstance(wfns, list):
                    dm = dm * weights[0]
                    for ex_wfn, w in zip(wfns[1:], weights[1:]):
                        dm = dm + w * np.tensordot(
                            ex_wfn[0], ex_wfn[0], axes=((-3, -2, -1), ) * 2)
                dm = self.add_dm_noise(dm, mpo, wfn, noise, forward)
                lsr = dm.tensor_svd(idx=3, pattern='++++++')
                l, _, _, error = pbalg.truncate_svd(
                    *lsr, cutoff=self.cutoff, max_bond_dim=bond_dim, eigen_values=True)
                wfn[:] = [l, np.tensordot(
                    l, wfn[0], axes=((0, 1, 2), ) * 2)]
            else:
                dm = np.tensordot(
                    wfn[0], wfn[0], axes=((0, 1, 2), ) * 2)
                if isinstance(wfns, list):
                    dm = dm * weights[0]
                    for ex_wfn, w in zip(wfns[1:], weights[1:]):
                        dm = dm + w * np.tensordot(
                            ex_wfn[0], ex_wfn[0], axes=((0, 1, 2), ) * 2)
                dm = self.add_dm_noise(dm, mpo, wfn, noise, forward)
                lsr = dm.tensor_svd(idx=3, pattern='-+--+-')
                _, _, r, error = pbalg.truncate_svd(
                    *lsr, cutoff=self.cutoff, max_bond_dim=bond_dim, eigen_values=True)
                wfn[:] = [np.tensordot(
                    wfn[0], r, axes=((-3, -2, -1), ) * 2), r]
        elif self.decomp_type == DecompositionTypes.SVD:
            assert not isinstance(wfns, list)
            wfn = self.add_wfn_noise(
                wfn[0], noise, forward)
            lsr = wfn.tensor_svd(idx=3, pattern='+++-+-')
            l, s, r, error = pbalg.truncate_svd(
                *lsr, cutoff=self.cutoff, max_bond_dim=bond_dim)
            if noise == 0:
                if forward:
                    wfn[:] = [l, np.tensordot(
                        s.diag(), r, axes=1)]
                else:
                    wfn[:] = [np.tensordot(
                        l, s.diag(), axes=1), r]
            else:
                if forward:
                    wfn[:] = [l, np.tensordot(
                        l, wfn[0], axes=((0, 1, 2), ) * 2)]
                else:
                    wfn[:] = [np.tensordot(
                        wfn[0], r, axes=((-3, -2, -1), ) * 2), r]
        else:
            assert False
        return error
