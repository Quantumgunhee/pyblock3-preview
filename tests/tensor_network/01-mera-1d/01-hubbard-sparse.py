
# 1D NN Hubbard model with 4 sites
# implemented using non-fermionic sparse tensors
# E dmrg = -2.875942808996

import numpy as np
import quimb.tensor as qtn
import warnings
import torch
import time
from pyblock3.hamiltonian import Hamiltonian
from pyblock3.fcidump import FCIDUMP

torch.autograd.set_detect_anomaly(False)
torch.set_num_threads(28)

import pyblock3.algebra.ad
pyblock3.algebra.ad.ENABLE_JAX = True
pyblock3.algebra.ad.ENABLE_AUTORAY = True

from pyblock3.algebra.ad import core
core.ENABLE_FUSED_IMPLS = True

from pyblock3.algebra.ad.core import SparseTensor, FermionTensor

warnings.filterwarnings(action='ignore', category=torch.jit.TracerWarning)

import autoray as ar

SparseTensor.shape = property(lambda x: tuple(ix.n_bonds for ix in x.infos))
FermionTensor.shape = property(lambda x: tuple(ix.n_bonds for ix in x.infos))
ar.register_function('pyblock3', 'array', lambda x: x)
ar.register_function('pyblock3', 'to_numpy', lambda x: x)
ar.register_function('torch', 'conjugate', torch.conj)

def torch_split_wrapper(old_fn):
    def new_fn(ary, indices_or_sections, *args, **kwargs):
        if len(indices_or_sections) == 0:
            return [ary]
        else:
            return old_fn(ary, indices_or_sections, *args, **kwargs)
    return new_fn

ar.register_function('torch', 'split', torch_split_wrapper, wrap=True)

L = 4
D = 20

np.random.seed(1234)

print(core.ENABLE_FUSED_IMPLS)

fd = FCIDUMP(pg='d2h', n_sites=L, n_elec=L, twos=0, ipg=0, orb_sym=[0] * L)
hamil = Hamiltonian(fd, flat=False)

class MERAState:

    @staticmethod
    def get_disentanglers_and_isometries(basis, target, low=0, high=1, dtype=float, max_bond=None):
        tensors_uni = []
        tensors_iso = []
        tensors_cap = []
        while len(basis) > 1:
            n_sites = len(basis) // 2
            uni_basis = []
            for i in range(n_sites):
                info_high = basis[i * 2] | basis[i * 2 + 1]
                if max_bond is not None:
                    info_high.truncate_no_keep(max_bond)
                ts = SparseTensor.random(
                    (basis[i * 2], basis[i * 2 + 1], info_high, info_high),
                    dtype=dtype, pattern="++--") * (high - low) + low
                tensors_uni.append(ts)
                uni_basis.append(info_high)

            iso_basis = []
            for i in range(n_sites):
                info_high = uni_basis[i] * uni_basis[(i + 1) % n_sites]
                if max_bond is not None:
                    info_high.truncate_no_keep(max_bond)
                if n_sites == 1:
                    info_high = info_high.filter(info_high.__class__({target : 1}))
                ts = SparseTensor.random(
                    (uni_basis[i], uni_basis[(i + 1) % n_sites], info_high),
                    dtype=dtype, pattern="++-") * (high - low) + low
                if n_sites == 1:
                    tensors_cap.append(ts)
                else:
                    tensors_iso.append(ts)
                iso_basis.append(info_high)

            basis = iso_basis

        return tensors_uni, tensors_iso, tensors_cap

    @staticmethod
    def unitize_(mera):
        for t in mera:
            if t.data.ndim - len(t.left_inds) == 1:
                new_data = t.data.left_canonicalize()[0]
            elif t.data.ndim - len(t.left_inds) == 2:
                fuse_info = t.data.kron_sum_info(2, 3, pattern=t.data.pattern[2:])
                new_data = t.data.fuse(2, 3, info=fuse_info).left_canonicalize()[0].unfuse(2, info=fuse_info)
                new_data.pattern = t.data.pattern
            else:
                assert False
            t.modify(data=new_data)
        return mera

    @staticmethod
    def c_operator(orb_sym, sz, basis, isite, dtype=float):
        ipg = orb_sym[isite]
        Q = basis[isite].symm_class
        repr = FermionTensor.zeros(bond_infos=(basis[isite], basis[isite]),
            dtype=dtype, dq=Q(1, sz, ipg), pattern="+-")
        repr.odd[Q(0, 0, 0)].data[0, 0] = 1
        repr.odd[Q(1, -sz, ipg)].data[0, 0] = sz
        return repr

    @staticmethod
    def d_operator(orb_sym, sz, basis, isite, dtype=float):
        ipg = orb_sym[isite]
        Q = basis[isite].symm_class
        repr = FermionTensor.zeros(bond_infos=(basis[isite], basis[isite]),
            dtype=dtype, dq=Q(-1, -sz, ipg), pattern="+-")
        repr.odd[Q(1, sz, ipg)].data[0, 0] = 1
        repr.odd[Q(2, 0, 0)].data[0, 0] = sz
        return repr

    @staticmethod
    def u_operator(basis, isite, dtype=float):
        Q = basis[isite].symm_class
        repr = FermionTensor.zeros(bond_infos=(basis[isite], basis[isite]),
            dtype=dtype, dq=Q(0, 0, 0), pattern="+-")
        repr.even[Q(2, 0, 0)].data[0, 0] = 1
        return repr

    @staticmethod
    def get_hamil_terms(topology, oyb_sym, basis, dtype=float):
        terms = []
        ca_op = [None] * len(basis)
        da_op = [None] * len(basis)
        cb_op = [None] * len(basis)
        db_op = [None] * len(basis)
        u_op = [None] * len(basis)
        for m in range(len(basis)):
            ca_op[m] = MERAState.c_operator(oyb_sym, 1, basis, m, dtype=dtype)
            da_op[m] = MERAState.d_operator(oyb_sym, 1, basis, m, dtype=dtype)
            cb_op[m] = MERAState.c_operator(oyb_sym, -1, basis, m, dtype=dtype)
            db_op[m] = MERAState.d_operator(oyb_sym, -1, basis, m, dtype=dtype)
            u_op[m] = MERAState.u_operator(basis, m, dtype=dtype)

        for (i, j, v) in topology:
            if i != j:
                ts = sum([
                    np.einsum('ij,kl->ikjl', v * ca_op[i], da_op[j]),
                    np.einsum('ij,kl->ikjl', -v * da_op[i], ca_op[j]),
                    np.einsum('ij,kl->ikjl', v * cb_op[i], db_op[j]),
                    np.einsum('ij,kl->ikjl', -v * db_op[i], cb_op[j])
                ])
                terms.append(((i, j), ts))
            else:
                ts = v * u_op[i]
                terms.append(((i, ), ts))

        return terms


class TensorBuilder:
    def __init__(self, hamil, max_bond=None):
        self.uni, self.iso, self.cap = MERAState.get_disentanglers_and_isometries(hamil.basis, hamil.target, max_bond=max_bond)
        self.i_uni = 0
        self.i_iso = 0
        self.i_cap = 0

    @property
    def uni_fill_fn(self):
        def uni_fill_fn(shape):
            uni = self.uni[self.i_uni]
            self.i_uni += 1
            print('uni ', uni.shape, shape)
            assert uni.shape == shape
            return uni
        return uni_fill_fn

    @property
    def iso_fill_fn(self):
        def iso_fill_fn(shape):
            iso = self.iso[self.i_iso]
            self.i_iso += 1
            print('iso ', iso.shape, shape)
            # assert iso.shape == shape
            return iso
        return iso_fill_fn

    @property
    def cap_fill_fn(self):
        def cap_fill_fn(shape):
            cap = self.cap[self.i_cap]
            self.i_cap += 1
            print('cap ', cap.shape, shape)
            # assert cap.shape == shape
            return cap
        return cap_fill_fn

def get_params(self):
    from jax.tree_util import tree_flatten
    params, tree = tree_flatten(self)
    self._tree = tree
    return params

def set_params(self, params):
    from jax.tree_util import tree_unflatten
    x = tree_unflatten(self._tree, params)
    self.blocks = x.blocks
    self.pattern = x.pattern

SparseTensor.get_params = get_params
SparseTensor.set_params = set_params


def f_set_params(self, params):
    from jax.tree_util import tree_unflatten
    x = tree_unflatten(self._tree, params)
    self.odd = x.odd
    self.even = x.even

FermionTensor.get_params = get_params
FermionTensor.set_params = f_set_params

tensor_builder = TensorBuilder(hamil, max_bond=D)

from quimb.experimental.merabuilder import MERA

mera = MERA.from_fill_fn(
    fill_fn=None,
    uni_fill_fn=tensor_builder.uni_fill_fn,
    iso_fill_fn=tensor_builder.iso_fill_fn,
    cap_fill_fn=tensor_builder.cap_fill_fn,
    L=L,         # number of sites, (doesn't have to be a power of block_size)
    D=D,           # max bond dimension
    block_size=2,
    phys_dim=4,
    cyclic=True,  # OBC
)

MERAState.unitize_(mera)

t = 1
u = 2
topology = [(i, i + 1, t) for i in range(L - 1)]
topology += [(i, i, u) for i in range(L)]

h_terms = MERAState.get_hamil_terms(topology, hamil.orb_sym, hamil.basis)
terms = { k : v for k, v in h_terms }

def norm_fn(mera):
    return MERAState.unitize_(mera.copy())

def local_expectation(mera, terms, where, optimize='auto-hq'):
    tags = [mera.site_tag(coo) for coo in where]
    mera_ij = mera.select(tags, 'any')
    mera_ij_G = mera_ij.gate(terms[where], where)
    mera_ij_ex = (mera_ij_G & mera_ij.H)
    return mera_ij_ex.contract(all, optimize=optimize)

def loss_fn(mera, terms, **kwargs):
    with qtn.contraction.contract_backend("numpy"):
        r = sum(
            local_expectation(mera, terms, where, **kwargs)
            for where in terms
        )
        return r

mera = norm_fn(mera)
print(loss_fn(mera, terms, optimize='auto-hq'))

def _tn_callback(self):
    if not hasattr(self, 'tt'):
        self.it = 0
        self.tt = time.perf_counter()
    self.loss_best = min(self.loss_best, self.loss)
    tt = time.perf_counter() - self.tt
    self.tt = time.perf_counter()
    self.it += 1
    msg = f"{self.it:5d} {self.loss:+.12f} [best: {self.loss_best:+.12f}] T = {tt:10.3f}"
    print(msg)

tnopt = qtn.TNOptimizer(
    mera,
    loss_fn=loss_fn,
    norm_fn=norm_fn,
    loss_constants={'terms': terms},
    loss_kwargs={'optimize': 'auto-hq'},
    autodiff_backend='torch',
    callback=_tn_callback,
    progbar=False,
    jit_fn=False,
)

tt = time.perf_counter()
tnopt.optimizer = 'l-bfgs-b'
tnopt.optimize(500)
# tnopt.optimizer = 'adam'
# mera_opt = tnopt.optimize(1000)

print(min(tnopt.losses))
print('time = ', time.perf_counter() - tt)
