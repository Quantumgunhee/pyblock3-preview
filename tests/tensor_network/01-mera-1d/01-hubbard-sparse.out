True
uni  (4, 4, 4, 4) (4, 4, 4, 4)
uni  (4, 4, 4, 4) (4, 4, 4, 4)
iso  (4, 4, 16) (4, 4, 16)
iso  (4, 4, 16) (4, 4, 16)
cap  (16, 16, 1) (16, 16)
1.9843658771366193
    1 +1.984365877137 [best: +1.984365877137] T =      0.000
    2 -0.608047495922 [best: -0.608047495922] T =      0.818
    3 -1.782915001664 [best: -1.782915001664] T =      0.843
    4 -2.502027414445 [best: -2.502027414445] T =      0.805
    5 -2.724752418929 [best: -2.724752418929] T =      0.814
    6 -2.811825853406 [best: -2.811825853406] T =      0.741
    7 -2.845265467291 [best: -2.845265467291] T =      0.819
    8 -2.861733746073 [best: -2.861733746073] T =      0.818
    9 -2.870139729255 [best: -2.870139729255] T =      0.820
   10 -2.871264624303 [best: -2.871264624303] T =      0.796
   11 -2.875514438391 [best: -2.875514438391] T =      0.798
   12 -2.875721152110 [best: -2.875721152110] T =      0.812
   13 -2.875894513748 [best: -2.875894513748] T =      0.806
   14 -2.875891799175 [best: -2.875894513748] T =      0.805
   15 -2.875915846696 [best: -2.875915846696] T =      0.788
   16 -2.875932345474 [best: -2.875932345474] T =      0.804
   17 -2.875937627807 [best: -2.875937627807] T =      0.771
   18 -2.875940459966 [best: -2.875940459966] T =      0.809
   19 -2.875941722875 [best: -2.875941722875] T =      0.769
   20 -2.875942518101 [best: -2.875942518101] T =      0.798
   21 -2.875942721029 [best: -2.875942721029] T =      0.813
   22 -2.875942790726 [best: -2.875942790726] T =      0.814
   23 -2.875942789892 [best: -2.875942790726] T =      0.765
   24 -2.875942802872 [best: -2.875942802872] T =      0.818
   25 -2.875942807099 [best: -2.875942807099] T =      0.811
-2.8759428070988107
time =  20.23057335987687
